Add python scripts and input json files

bquan0 opened the pull request:
Adds the following files:
* `input.json` specifies .stl files and their respective textures, as well as other information required by `stl_to_tscn.py`
* `input_textures.json` allows contributors to use extra textures that we don't provide in our library.
* `mcnp_to_csv` an old script that converts MCNP .txt data into a .csv file that can be parsed by the in game Godot script and loaded into an array to be used for radiation data sampling. 
* `stl_to_tscn.py` generates .obj meshes, .glb collision scenes, and the .tscn file specifying the textures of each mesh and the scene tree. Turns .stl files into a Godot scene
* `textures.py` helper module of `stl_to_tscn.py`. Contains various texture information such as .jpg file names and the texture's folder name.

bquan0 committed:
add python scripts and input json files

MicahGale commented on mcnp_to_csv.py:
```diff
@@ -0,0 +1,60 @@
+# make sure the user entered the mcnp filename and the output filename as CLAs
+if len(sys.argv) != 3:
+    print("usage: python3 mcnp_to_csv.py <mcnp filename> <output .csv filename>")
+    quit()
```
Switch to using `argparse`

MicahGale commented on mcnp_to_csv.py:
```diff
@@ -0,0 +1,60 @@
+
+# open files
+mcnpFile = open(sys.argv[1], 'r')
+outputFile = open(output_filename, 'w')
```
Use a with block.

connoramoreno commented on mcnp_to_csv.py:
```diff
@@ -0,0 +1,60 @@
+    quit()
+
+# append output filename with .csv if it doesn't end in .csv
+output_filename = sys.argv[2]
+if output_filename[-4:] != ".csv":
+    output_filename = output_filename + ".csv"
```
Look into using `os.path` or `pathlib.Path` to extract the file extension

MicahGale commented on mcnp_to_csv.py:
```diff
@@ -0,0 +1,60 @@
+# get array of boundaries for x,y,z
+# 1a. find line that has x boundaries
+mcnpLine = mcnpFile.readline()
+while mcnpLine.find("X direction:") == -1:
```
`while "X direction:" in mcnpline:` is cleaner

connoramoreno commented:
Generally, we like to use the [PEP8](https://peps.python.org/pep-0008/) style for standardized code formatting. You can change your formatting manually or use a utility like [Black](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html) to auto-format your code (I believe there's a VS Code extension for Black but I have never used it).

MicahGale commented on stl_to_tscn.py:
```diff
@@ -0,0 +1,245 @@
+        elif uv_map_type == "sphere":
+            bpy.ops.uv.sphere_project()
+        elif uv_map_type == "unwrap":
+            bpy.ops.uv.unwrap()
```
Try replacing this with a constant dict. Like `OPERATOR_MAP = {"smart": bpy.ops.uv.smart_project...}`

MicahGale replied:
May want to wrap call inside a 
``` python
try:

except KeyError as e:
   raise ... from e
```

MicahGale commented on stl_to_tscn.py:
```diff
@@ -0,0 +1,245 @@
+    bpy.context.view_layer.objects.active = active_object
+
+    # Export to obj
+    bpy.ops.wm.obj_export(filepath = output_prefix + ".obj")
```
f strings!

connoramoreno replied:
Example f-string: `filepath = f"{output_prefix}.obj"`

MicahGale commented on stl_to_tscn.py:
```diff
@@ -0,0 +1,245 @@
+
+    # rename all objects as -colonly
+    num_objs = len(bpy.data.objects)
+    for k in range(num_objs):
```
`for k, _ in enumerate()`:

MicahGale replied:
`object_number`, etc would be a better index.

MicahGale commented on stl_to_tscn.py:
```diff
@@ -0,0 +1,245 @@
+# argparse to take in input.txt
+parser = argparse.ArgumentParser(prog = ".stl to .obj and .tscn file converter")
+parser.add_argument("input_file", type = str, help = "input json filename")
+args = parser.parse_args()
```
Create a `main` function.

then add:
``` python
if __name__ == "__main__":
   main()
```

MicahGale commented on stl_to_tscn.py:
```diff
@@ -0,0 +1,245 @@
+    output_folder = header["output_folder"]
+    try:
+        os.mkdir("./" + output_folder)
+    except OSError as error:
+        print("NOTICE: " + output_folder + " folder already exists")
```
have an "if exists" guard, and don't catch without re raising here.

MicahGale replied:
Actually don't have a `try` block

MicahGale commented on stl_to_tscn.py:
```diff
@@ -0,0 +1,245 @@
+                    ext_resource += "[ext_resource path=\"res://" + Textures.texture_folder + "/" + Textures.texture_dict[texture].folder \
+                                    + "/" + value + "\" type=\"Texture\" id=" + str(ext_resource_id) + "]\n"
+                    # add to sub_resource
+                    if jpg == "albedo":
+                        sub_resource += "albedo_texture = ExtResource( " + str(ext_resource_id) + " )\n"
+                    elif jpg == "roughness":
+                        sub_resource += "roughness_texture = ExtResource( " + str(ext_resource_id) + " )\n"
```
``` python
if jpg in ["albedo", "roughness"]:
    sub_resource += f"{jpg}_text = ExtResource( {ext_resource_id} )\n"
```

MicahGale commented on stl_to_tscn.py:
```diff
@@ -0,0 +1,245 @@
+        stl_file = m["stl_file"]
+
+        # output filename without a file extension
+        output_prefix = output_folder + "/" + stl_file[:-4]
```
`os.path.splitext`

MicahGale commented on stl_to_tscn.py:
```diff
@@ -0,0 +1,245 @@
+        # Generate .obj file
+        generate_obj_file(stl_file, output_prefix, m["uv_map"])
+        # add it to ext_resource
+        ext_resource += "[ext_resource path=\"res://models/" + output_prefix + \
```
Switch to single quote so you don't need to escape: `f'..."res://"...'`

MicahGale commented on stl_to_tscn.py:
```diff
@@ -0,0 +1,245 @@
+        ext_resource += "[ext_resource path=\"res://models/" + output_prefix + \
+                        ".obj\" type=\"ArrayMesh\" id=" + str(ext_resource_id) + "]\n"
+        # add it to nodes
+        nodes += "[node name=\"" + stl_file[:-4] + "\" type=\"MeshInstance\" parent=\".\"]\nmesh = ExtResource( " \
```
make an base_name variable instead of `stl_file[:-4]`

MicahGale commented on stl_to_tscn.py:
```diff
@@ -0,0 +1,245 @@
+        #######################
+        # Generate collisions #
+        #######################
+        if m["collisions"]:
+            generate_collisons(stl_file, output_prefix)
+            # add it to ext_resource
+            ext_resource += "[ext_resource path=\"res://models/" + output_prefix + \
+                            ".glb\" type=\"PackedScene\" id=" + str(ext_resource_id) + "]\n"
+                        
+            nodes += "[node name=\"" + stl_file[:-4] + "Col\" parent=\".\" instance=ExtResource( " + str(ext_resource_id) + " )]\n\n"
+
+            # move on to next ext_resource
+            ext_resource_id = ext_resource_id + 1
```
functionalize.

bquan0 replied:
`generate_collisions` is already a function that generates the `.glb` collision files, and it would be awkward to put the other lines here into the function because that would require passing in the `ext_resource` and `nodes` string into the function.

MicahGale commented on textures.py:
```diff
@@ -0,0 +1,90 @@
+    uv1_scale = "Vector3( 1.75, 2.25, 1 )"
+)
+
+# TODO put in different .json file
```
What's the resolution on this?

MicahGale commented on textures.py:
```diff
@@ -0,0 +1,90 @@
+import json
+
+# filenames of all the PBR textures included in game in the textures folder, stored as a dict for each file
+texture_folder = "textures"
+
+normal_options = "normal_enabled = true\nnormal_scale = -1.0\n"
+depth_options = "depth_enabled = true\ndepth_scale = -0.01\ndepth_deep_parallax = false\ndepth_flip_tangent = false\ndepth_flip_binormal = false\n"
```
constants should be all caps snake case.

MicahGale commented on textures.py:
```diff
@@ -0,0 +1,90 @@
+        self.uv1_scale = uv1_scale 
+
+
+ceiling_tile = Texture(
+    folder = "CeilingTile",
+    jpg_dict = {
+        "albedo" : "OfficeCeiling001_1K-JPG_Color.jpg",
+        "roughness" : "OfficeCeiling001_1K-JPG_Roughness.jpg",
+        "normal" : "OfficeCeiling001_1K-JPG_NormalGL.jpg", 
+        "depth" : "OfficeCeiling001_1K-JPG_Displacement.jpg"
+    },
```
To reiterate use a json file here. Then make a classmethod to generate a new texture

``` python
@classmethod
def make_texture_from_json(json_obj): 
     return Texture(**json_obj)
```

MicahGale reviewed (commented):
This is a good start:

High level:

1. More functions
2. use f-strings
3. Reduce the need for if else

bquan0 committed:
apply recommendations

bquan0 commented:
* Added a `.gitignore` because I moved the repo into the folder where I was actually working on these files, and that folder has a lot of extra files from previous and current tests that don't need to be added to this repo. This also caused the addition of the `stl_to_obj` folder. 
* Applied the f string and os.path.splitext recommendations.
* Renamed `input_textures.json` to `extra_textures.json`. Added `textures.json` to replace the hardcoded textures I previously had in `textures.py`. 
* Added an option in `stl_to_tscn.py` to disable limited dissolve (since that was causing issues with the stellarator models) as well as add a metallic texture option (same as albedo and roughness in terms of formatting).
* Added a `generate_texture()` function in stl_to_tscn.py to make the part where we generate textures cleaner.

MicahGale commented:
A general comment. I see that you only have two commits for this pull request. In general this is far too few. I know I am guilty of doing very small (I like to think atomic) commits, but I think generally each task should be it's own task. For example switching to f strings, switching to a with based file handle, etc. Here's a helpful blog on it: https://www.freecodecamp.org/news/how-to-write-better-git-commit-messages/

gonuke commented on .gitignore:
```diff
@@ -0,0 +1,16 @@
+*.csv
+*.stl
+# python scripts used for testing purposes
+stl_to_obj/bpy_test.py
```
I think you should work to have your files laid out in a way that this doesn't matter.  In this case, you may actually want to include the files you use for testing in the repo, assuming they are good demonstrations and/or tests

bquan0 replied:
All the file extensions should be a good generalization for all users, because they are large input/output files that we don't want in the repo. As for the extra python scripts and json files that I've included, we can either continue with how it is, or bundle them all into one folder (which just pushes off the problem even further since that entire folder won't be committed into the repo).

MicahGale replied:
@gonuke this `bpy_test.py` file is more a "sandbox" than a testing suite. I think for now this is fine, but maybe by next PR it's probably time to move beyond sandboxing and switching to something like `pytest`?

bquan0 replied:
Just cleaned up the `.gitignore` and removed all the personal files. I'm now using `.git/info/exclude" to ignore those files.

gonuke commented on mcnp_to_csv.py:
```diff
@@ -0,0 +1,57 @@
+    # 1b. split line into all the boundaries, write to file with comma separation
+    # do this 3 times for the 3 axes
+    for a in range(3):
+        boundariesArray = mcnpLine.split()
+        outputStr = ""
+        ba_length = len(boundariesArray)
+        for index, boundary in enumerate(boundariesArray):
+            if index >= 2 and index < ba_length-1:
+                outputStr += boundary + ","
+            elif index == ba_length-1:
+                outputStr += boundary + "\n"
```
Use python's standard built-in string manipulation!
```suggestion
        outputStr = ','.join(mcnpLine.split().strip()) + '\n'
```

gonuke commented on mcnp_to_csv.py:
```diff
@@ -0,0 +1,57 @@
+        mcnpLine = mcnpFile.readline()
+    # 1b. split line into all the boundaries, write to file with comma separation
+    # do this 3 times for the 3 axes
+    for a in range(3):
```
Readability
```suggestion
    num_dims = 3
    for dim in range(num_dims):
```

gonuke commented on mcnp_to_csv.py:
```diff
@@ -0,0 +1,57 @@
+    mcnpLine = mcnpFile.readline() # this will have the first centroid (x,y,z) and result
+    # iterate through all voxels
+    while mcnpLine:
+        voxelArray = mcnpLine.split()
+        outputStr = ""
+        va_length = len(voxelArray)
+        # write all data of a voxel with comma separation
+        for index, data in enumerate(voxelArray):
+            if index >= 1 and index < va_length-1:
+                outputStr += data + ","
+            elif index == va_length-1:
+                outputStr += data + "\n"
```
```suggestion
        outputStr = ','.join(mcnpLine.split().strip()) + '\n'
```

gonuke commented on mcnp_to_csv.py:
```diff
@@ -0,0 +1,57 @@
+        outputFile.write(outputStr)
+        mcnpLine = mcnpFile.readline()
+
+
```
Add an `if __name__ == "__main__":` block and put everything in funcionts.

gonuke commented on stl_to_obj/textures.py:
```diff
@@ -0,0 +1,57 @@
+        # load JSON object as dictionary
+        data = json.load(f)
+
+        textures = data["textures"]
+        
+        for t in textures:
+            # create Texture
+            folder = t["folder"]
+            jpg_dict = t["jpg_dict"]
+            uv1_scale = ""
+            if "uv1_scale" in t:
+                uv1_scale = t["uv1_scale"]
+            texture = Texture(
+                folder = folder,
+                jpg_dict = jpg_dict,
+                uv1_scale = uv1_scale
+            )
+
+            # add it to texture_dict
+            texture_dict[folder] = texture
```
```suggestion
        textures = data["textures"]
        
        for texture_data in data["textures"]:
            if 'uv_scale1' not in texture_data:
                texture_data['uv_scale1'] = ""
            texture_dict[texture_data['folder']] = Texture(texture_data['folder'],
                                                           texture_data['jpg_dict'],
                                                           texture_data['uv_scale1'])
```

gonuke commented on stl_to_obj/textures.py:
```diff
@@ -0,0 +1,57 @@
+
+# add textures from a .json file to dict 
+# allows contributors to use their own textures (that we don't have)
+def add_textures(filename):
```
```suggestion
def load_textures(filename):
```

gonuke commented on stl_to_obj/textures.py:
```diff
@@ -0,0 +1,57 @@
+    with open(filename, 'r') as f:
+        # load JSON object as dictionary
+        data = json.load(f)
+
```
I think everything after this can be unindented - the file is loaded in the single line above and `f` is no longer needed.

gonuke commented on stl_to_obj/textures.py:
```diff
@@ -0,0 +1,57 @@
+# debug printout to check if we load our provided textures correctly
+if __name__ == "__main__":
+    add_textures("textures.json")
+    for i in texture_dict:
+        print(i)
+
+    for i in other_textures:
+        print(i)
```
Add an option to your `argparse` to list the textures and otherwise don't print any output.

bquan0 replied:
I don't think this is necessary, because usually this module is used by other python scripts like `stl_to_tscn.py` instead of being run itself. Adding the argparse allows us to hide the debug printout, but when we run this module itself and we don't see any debug printout, it appears like the module did nothing. 

Maybe I should just remove the `if __name__ == "__main__":` section then?

gonuke commented on stl_to_obj/textures.json:
```diff
@@ -0,0 +1,38 @@
+        }
+        
+    ],
+    "other_textures": {
+        "radioactive": "albedo_color = Color( 0, 1, 0.156863, 1 )\nmetallic = 0.3\nroughness = 0.0\n",
+        "glass": "flags_transparent = true\nalbedo_color = Color( 0.364706, 0.364706, 0.364706, 0.505882 )\nmetallic = 1.0\nroughness = 0.0\n"
+    }
```
Why are these other textures defined differently?  Should they be called something different than textures?  Can we generalize this at all?

bquan0 replied:
These textures are defined differently because they don't require any .jpgs, and therefore their information can just be copied directly to the sub_resource. It's easier to have them as separate strings rather than trying to generalize them to fit in with the other textures that require .jpgs

gonuke commented on stl_to_obj/stl_to_tscn.py:
```diff
@@ -0,0 +1,259 @@
+        # parse json file #
+        ###################
+        # load JSON object as dictionary
+        data = json.load(f)
```
Move this to be just after the `with open()` line and then unindent everything else - the file is 100% loaded after this one line and the `f` is long longer needed

gonuke commented on stl_to_obj/stl_to_tscn.py:
```diff
@@ -0,0 +1,259 @@
+
+        input_folder = header["input_folder"]
+
+        # make the output folder
+        output_folder = header["output_folder"]
+        try:
+            os.mkdir(f"./{output_folder}")
+        # delete the folder and remake it if it does exist
+        except OSError as error:
+            print(f"NOTICE: {output_folder} folder already exists")
+            shutil.rmtree(output_folder) # delete folder and contents
+            os.mkdir(f"./{output_folder}")
```
put this in a function, and possibly add an argparse option to force this delete or not

bquan0 replied:
The delete always happens if the folder exists because we want to do a complete regeneration of the files. We don't want to write to a folder that exists and has files in it because then there could be extra files leftover that don't contribute to the .tscn file that is generated.

gonuke commented on stl_to_obj/stl_to_tscn.py:
```diff
@@ -0,0 +1,259 @@
+        # load provided textures
+        Textures.add_textures("textures.json")
+        # (optional) load extra textures 
+        if "extra_textures" in header: 
+            Textures.add_textures(header["extra_textures"])
```
This syntax is a little confusing since you have an `extra_textures` dictionary that is different from the `extra_textures` file...

bquan0 replied:
I will rename `extra_textures.json` to `more_textures.json`

gonuke commented on stl_to_obj/stl_to_tscn.py:
```diff
@@ -0,0 +1,259 @@
+        sub_resource_id = 1
+
+        # MeshInstances and collision scenes (.glb)
+        nodes = ""
```
You don't need this since the first time you add something to it always happens - just initialize with the first value instead of adding to it.

gonuke commented on stl_to_obj/stl_to_tscn.py:
```diff
@@ -0,0 +1,259 @@
+        texture_index = dict()
+
+        for m in meshes:
+            ####################
+            # generate texture #
+            ####################
+            texture = m["texture"]
+
+            # don't generate textures more than once
+            if texture not in texture_index:
+                ext_res_addon, ext_resource_id, sub_res_addon = generate_texture(texture, ext_resource_id, sub_resource_id)
+                ext_resource += ext_res_addon
+                sub_resource += f"{sub_res_addon}\n"
+                    
+                # add the sub resource's index to texture_index so we can use it later 
+                texture_index[texture] = sub_resource_id
+                sub_resource_id += 1 
```
Do this in a function - with object oriented, this could be a member function with knowledge of the right data

gonuke commented on stl_to_obj/stl_to_tscn.py:
```diff
@@ -0,0 +1,259 @@
+                sub_resource_id += 1 
+
+
+            #####################
+            # generate obj mesh #
+            #####################
+            stl_file = m["stl_file"]
+            stl_file_extless, _ = os.path.splitext(stl_file)
+            stl_file_location = f"{input_folder}/{stl_file}"
+
+            # output filename without a file extension
+            output_prefix = output_folder + "/" + stl_file_extless
+            
+            # Generate .obj file
+            if "limited_dissolve" in m:
+                generate_obj_file(stl_file_location, output_prefix, m["uv_map"], limited_dissolve=False)
+            else:
+                generate_obj_file(stl_file_location, output_prefix, m["uv_map"])
+
+            # add it to ext_resource
+            ext_resource += f'[ext_resource path="res://models/{output_prefix}.obj" type="ArrayMesh" id={ext_resource_id}]\n'
+            # add it to nodes
+            nodes += f'[node name="{stl_file_extless}" type="MeshInstance" parent="."]\nmesh = ExtResource( {ext_resource_id} )\nmaterial/0 = SubResource( {texture_index[texture]} )\n\n'
+
+            # move on to next ext_resource
+            ext_resource_id = ext_resource_id + 1
```
Do this in a function - member function?

gonuke commented on stl_to_obj/stl_to_tscn.py:
```diff
@@ -0,0 +1,259 @@
+            # Generate collisions #
+            #######################
+            if m["collisions"]:
+                generate_collisons(stl_file_location, output_prefix)
+                # add it to ext_resource
+                ext_resource += f'[ext_resource path="res://models/{output_prefix}.glb" type="PackedScene" id={ext_resource_id}]\n'
+                            
+                nodes += f'[node name="{stl_file_extless}Col" parent="." instance=ExtResource( {ext_resource_id} )]\n\n'
+
+                # move on to next ext_resource
+                ext_resource_id = ext_resource_id + 1
```
if `generate_collisions` was a member function you could to all of this in that function?

Also - `generate_collisions` has return values that you don't use?

bquan0 replied:
`generate_collisions` doesn't return anything

gonuke replied:
I got mixed up with another function, I guess

gonuke commented on stl_to_obj/stl_to_tscn.py:
```diff
@@ -0,0 +1,259 @@
+            output_prefix = output_folder + "/" + stl_file_extless
+            
+            # Generate .obj file
+            if "limited_dissolve" in m:
+                generate_obj_file(stl_file_location, output_prefix, m["uv_map"], limited_dissolve=False)
+            else:
+                generate_obj_file(stl_file_location, output_prefix, m["uv_map"])
```
Confusing logic choice - the existing of `'limited_dissolve'` means that it's *False*?  Is that correct?
If so, then I suggest:
```suggestion
            generate_obj_file(stl_file_location, output_prefix, m["uv_map"], limited_dissolve=('limited dissolve' not in m))
```

bquan0 replied:
Yeah, if `limited_dissolve` exists then we don't run limited dissolve on the mesh. This is because the default is that we do apply limited dissolve.

gonuke commented on stl_to_obj/stl_to_tscn.py:
```diff
@@ -0,0 +1,259 @@
+    # check if it's a PBR texture (require .jpgs)
+    if texture in Textures.texture_dict:
+        sub_res_addon += f'[sub_resource type="SpatialMaterial" id={sub_resource_id}]\n'
+        for jpg in Textures.texture_dict[texture].jpg_dict:
+            value = Textures.texture_dict[texture].jpg_dict[jpg]
```
```suggestion
        for jpg, value in Textures.texture_dict[texture].jpg_dict.items():
```

gonuke reviewed (changes requested):
A lot of room for cleaner and more modular python here

bquan0 committed:
use python built-in string manipulation in mcnp_to_csv.py

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

bquan0 committed:
fix built-in string generations, make a mcnp_to_csv function

bquan0 committed:
rename function and simplify load_textures function code

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

bquan0 committed:
rename add_textures to load_textures everywhere, close file sooner in load_textures

gonuke commented on stl_to_obj/textures.py:
```diff
@@ -0,0 +1,48 @@
+# add textures from a .json file to dict 
+# allows contributors to use their own textures (that we don't have)
+def load_textures(filename):
+    data = ""
```
I don't think you need this since you initialize `data` with file data.

bquan0 replied:
You're right, this entire time I was under the impression that data would go out of scope after the with block if I didn't initialize it before the with block.

bquan0 committed:
streamline key value access of dict

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

bquan0 committed:
create TscnGen class

bquan0 commented:
I added the `TscnGen` class in `stl_to_tscn.py` and put all the texture, obj, collision generating methods in there. I guess it helps with readability since all the functionality has been put inside the functions instead of having some code before and after the old functions.

I also added a new parameter for each .stl file called `mesh_compression`, which has the options of `limited_dissolve`, `tris_to_quads`, and none. This was motivated by the fact that limited dissolve sometimes tried combining faces over holes. Tris to quads doesn't have this problem, but results in a larger .obj file (still smaller than not doing any mesh compression).

gonuke commented on stl_to_obj/stl_to_tscn.py:
```diff
@@ -0,0 +1,208 @@
+        # don't generate textures more than once
+        if texture not in self.texture_index:
+            # handle PBR textures (require .jpgs)
+            if texture in Textures.texture_dict:
+                self.sub_resource += f'[sub_resource type="SpatialMaterial" id={self.sub_resource_id}]\n'
```
More readable with temp variable `t_info` replacing `Textures.texture_dict[texture]` everywhere
```suggestion
            if texture in Textures.texture_dict:
                t_info = Textures.texture_dict[texture]
                self.sub_resource += f'[sub_resource type="SpatialMaterial" id={self.sub_resource_id}]\n'
```

gonuke commented on stl_to_obj/stl_to_tscn.py:
```diff
@@ -0,0 +1,208 @@
+        bpy.ops.import_scene.fbx(filepath = fbx_file)
+
+        # rename all objects as -colonly
+        for obj_index, _ in enumerate(bpy.data.objects):
```
```suggestion
        for obj_index in range(len(bpy.data.objects)):
```

gonuke commented on stl_to_obj/textures.py:
```diff
@@ -0,0 +1,47 @@
+
+    # load in textures that don't require .jpgs
+    if "other_textures" in data:
+        for key in data["other_textures"]:
+            other_textures[key] = data["other_textures"][key]
```
Isn't this just a copy?
```suggestion
        other_textures = data["other_textures"]
```

bquan0 replied:
This change doesn't work. I also tried using `.copy()` and `copy.deepcopy()` and it just doesn't work. The reason is because the value of `other_textures` gets set to null after `load_textures()` if we try to make a copy in this way

connoramoreno replied:
When you say `other_textures` gets set to null, do you mean an empty dictionary or `None`? If it's an empty dictionary, it might have to do with the fact that `other_textures` is instantiated globally as an empty dictionary.

I say this without having poked around in the code myself, though. Not sure if the issue is happening when `other_textures` is imported into `stl_to_tscn.py` or just within this script itself.

gonuke commented on stl_to_obj/textures.py:
```diff
@@ -0,0 +1,47 @@
+            texture_data['uv1_scale'] = ""
+        texture_dict[texture_data['folder']] = Texture(texture_data['folder'],
+                                                        texture_data['jpg_dict'],
+                                                        texture_data['uv1_scale'])
```
With `dict.get()` you can avoid the need to initialize all of these in the loop before this
```suggestion
                                                        texture_data.get('uv1_scale', ""))
```

gonuke reviewed (changes requested):
I have a few more suggestions to help readability

bquan0 committed:
add recommendations for readability

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

bquan0 committed:
fix suggestions

bquan0 committed:
clean up .gitignore

gonuke reviewed (approved):
Thanks @bquan0 - this looks good to me

gonuke commented:
Pinged @connoramoreno and @MicahGale to re-review following changes.

connoramoreno commented on mcnp_to_csv.py:
```diff
@@ -0,0 +1,48 @@
+def mcnp_to_csv(mcnpFile, outputFile):
+    # append output filename with .csv regardless of what its file extension is
+    output_filename, _ = os.path.splitext(outputFile)
+    output_filename += ".csv"
```
You might consider using `pathlib.Path` for all file-structure-manipulating needs. It's Python's more modern alternative to `os.path` that allows for platform-independence. It's built-in to Python and easy to use, once you get used to it. I'm happy to give some pointers for its use if you'd like. Feel free to look at ParaStell for an example of its use.

The caveat is that `pathlib.Path` requires Python 3.4+. Not sure if that's an issue here.

connoramoreno commented on mcnp_to_csv.py:
```diff
@@ -0,0 +1,48 @@
+        for dim in range(num_dims):
+            outputStr = ','.join(mcnpLine.strip().split()[2:]) + '\n'
+            outputFile.write(outputStr)
+            mcnpLine = mcnpFile.readline()
```
```suggestion
        dims = ['x', 'y', 'z']
        for dim in dims:
            outputStr = ','.join(mcnpLine.strip().split()[2:]) + '\n'
            outputFile.write(outputStr)
            mcnpLine = mcnpFile.readline()
```
Here's a somewhat arbitrary suggestion. It won't change the functionality at all but might make it even clearer what's actually happening in this block (useful when you or someone else returns to this script after some time away).

connoramoreno commented on stl_to_obj/stl_to_tscn.py:
```diff
@@ -0,0 +1,209 @@
+import bpy
+import json
+import shutil
+import textures as Textures # custom module created to store texture info
```
```suggestion
import argparse
import os
import json
import shutil

import aspose.threed as a3d
import bpy

import textures as Textures # custom module created to store texture info
```
You might consider grouping imports as suggested to follow [the corresponding PEP8 guidelines](https://peps.python.org/pep-0008/#imports).

For imports, guidelines are to group imports as follows, with whitespace between import groups:

1. Standard library imports (Python libraries that are built-in)
2. Related third party imports (dependencies)
3. Local application/library specific imports (custom modules housed locally)

connoramoreno replied:
You might reconsider renaming this package. The name `Textures` is easy to get confused with the class `Texture` since a leading capital letter is standard formatting for classes rather than packages (see relevant [PEP8 naming convention](https://peps.python.org/pep-0008/#package-and-module-names))

connoramoreno commented on stl_to_obj/stl_to_tscn.py:
```diff
@@ -0,0 +1,209 @@
+    "unwrap": bpy.ops.uv.unwrap
+}
+
+class TscnGen:    
```
I suggest creating [a docstring](https://peps.python.org/pep-0008/#documentation-strings) for this class and its methods (excluding `__init__`), as well as for functions outside this class, to clarify what this code's purpose is. It's extra work but, as mentioned elsewhere, is useful when you or others return to this code after a while away.

connoramoreno commented on stl_to_obj/stl_to_tscn.py:
```diff
@@ -0,0 +1,209 @@
+        self.texture_index = dict()
+
+    def add_texture(self, texture):
+        # don't generate textures more than once
+        if texture not in self.texture_index:
```
```suggestion
        # don't generate textures more than once
        if texture in self.texture_index:
            w = Warning(f'{texture} has already been generated.')
            raise w
        else:
```
You can consider raising some warning here instead to notify users that they are attempting to add the texture more than once.

bquan0 replied:
I don't think this will be an issue because it is possible that users will want to use the same texture on different `.stl` files. Not regenerating textures is primarily to avoid doing more work than we have to (if the texture has already been imported, we can just reuse it instead of generating it again).

connoramoreno commented on stl_to_obj/textures.py:
```diff
@@ -0,0 +1,46 @@
+NORMAL_OPTIONS = "normal_enabled = true\nnormal_scale = -1.0\n"
+DEPTH_OPTIONS = "depth_enabled = true\ndepth_scale = -0.01\ndepth_deep_parallax = false\ndepth_flip_tangent = false\ndepth_flip_binormal = false\n"
+
+class Texture:
```
Again, docstrings.

connoramoreno commented on stl_to_obj/textures.py:
```diff
@@ -0,0 +1,46 @@
+    # load in textures that don't require .jpgs
+    if "other_textures" in data:
+        for key in data["other_textures"]:
+            other_textures[key] = data["other_textures"][key]
```
```suggestion
        for key, value in data["other_textures"].items():
            other_textures[key] = value
```
I believe this could work as well, though you might consider renaming `key` and `value` to something like `name` and `texture`, respectively, to clarify what they are.

However, just copying the whole dictionary at once in some way like Paul suggested would be ideal.

connoramoreno commented on stl_to_obj/stl_to_tscn.py:
```diff
@@ -0,0 +1,209 @@
+        # reset bpy
+        bpy.ops.wm.read_factory_settings(use_empty=True)
+        # import .stl into bpy
+        bpy.ops.import_mesh.stl(filepath = f"{self.input_folder}/{stl_file_extless}.stl")
```
`pathlib.Path` would allow you to combine directory and file strings in a manner specific to any OS without having to introduce OS-specific syntax.
```suggestion
        import_path = Path(self.input_folder) / Path(stl_file_extless).with_suffix('.stl')
        bpy.ops.import_mesh.stl(filepath=str(import_path))
```

connoramoreno commented on stl_to_obj/stl_to_tscn.py:
```diff
@@ -0,0 +1,209 @@
+            # Call project operator to generate uv maps
+            try:
+                UV_MAPS[uv_map]()
+            except KeyError as e:
+                print(f"no uv map generated for {stl_file_extless} due to invalid uv_map")
```
```suggestion
            except KeyError as e:
                print(f"no uv map generated for {stl_file_extless} due to invalid uv_map")
                raise e
```
If you're going to store the `KeyError`, you may as well use it. Otherwise, you can avoid storing it by removing `as e` from the `except` line:
```suggestion
            except KeyError:
                e = KeyError(f"no uv map generated for {stl_file_extless} due to invalid uv_map")
                raise e
```
From my understanding, you generally want to avoid print statements when raising an exception, and instead want to raise the exception itself to make it clearer to the user what went wrong.

connoramoreno commented on stl_to_obj/stl_to_tscn.py:
```diff
@@ -0,0 +1,209 @@
+            obj.select_set(False)
+
+        # Export to obj
+        bpy.ops.wm.obj_export(filepath = f"{self.output_folder}/{stl_file_extless}.obj")
+        # remove the .mtl file
+        os.remove(f"{self.output_folder}/{stl_file_extless}.mtl")
```
More opportunities to use `pathlib` to allow platform-independent use. There are other places to use it in this code, but I'll avoid pointing them all out for brevity.
```suggestion
        export_path = Path(self.output_folder) / Path(stl_file_extless).with_suffix('.obj')
        bpy.ops.wm.obj_export(filepath = str(export_path))
        
        # remove the .mtl file
        mtl_filepath = Path(self.output_folder) / Path(stl_file_extless).with_suffix('.mtl')
        Path.unlink(mtl_filepath)
```

connoramoreno commented on stl_to_obj/stl_to_tscn.py:
```diff
@@ -0,0 +1,209 @@
+        bpy.ops.import_scene.fbx(filepath = fbx_file)
+
+        # rename all objects as -colonly
+        for obj_index in range(len(bpy.data.objects)):
+            # every re-name moves the object to the back of the list. 
+            bpy.data.objects[0].name = "Volume" + str(obj_index) + "-colonly"
```
```suggestion
        for obj_index, _ in enumerate(bpy.data.objects):
            # every re-name moves the object to the back of the list. 
            bpy.data.objects[0].name = "Volume" + str(obj_index) + "-colonly"
```
More Pythonic to use `enumerate(...)` than `range(len(...))`. The underscore for the second item returned by enumerate (each object in `bpy.data.objects`) will treat that item as a throw-away since it's not needed.

You could alternatively use
```suggestion
        for obj_index, obj in enumerate(bpy.data.objects):
            # every re-name moves the object to the back of the list. 
            obj.name = "Volume" + str(obj_index) + "-colonly"
```
but that will preserve the order of `bpy.data.objects`. I'll take your word for it that each renaming moves the corresponding object to the back of the list rather than just rename the object in the first element of the list, while leaving the others unaffected.

bquan0 replied:
I applied that change in [this commit](https://github.com/cnerg/level-gen/pull/1/commits/8130e5b253c059ff030b6e482e59aec40f82f188) previously, but then @gonuke recommended the `range(len(bpy.data.objects))`.

gonuke replied:
I liked the version with `range()` better than the version with `index, _ in enumerate()` because we don't use the `index` to access the entry in the list.  It seemed confusing to get an index and then reference the 0 element.

If the version with `index, obj in enumerate()` works - not sure because of the way we alter the list as part of this loop - then it's probably better.

connoramoreno commented on stl_to_obj/stl_to_tscn.py:
```diff
@@ -0,0 +1,209 @@
+
+
+def make_folder(folder):
+    try:
+        os.mkdir(f"./{folder}")
+    # delete the folder and remake it if it does exist
+    except OSError as error:
+        print(f"NOTICE: {folder} folder already exists, so it was deleted and remade")
+        shutil.rmtree(folder) # delete folder and contents
+        os.mkdir(f"./{folder}")
```
Opportunity to implement better practices here for `try`, `except` blocks as mentioned above. Also note that raising an exception here will terminate the process. If that's not the intended functionality, try raising a warning instead.

connoramoreno reviewed (commented):
Nice work here @bquan0! I've made some suggestions generally related to what I think are coding best practices, and use of `pathlib.Path` instead of `os` for platform-independence if that's desirable.

bquan0 committed:
Apply Connor's suggestions

Co-authored-by: connoramoreno <67344181+connoramoreno@users.noreply.github.com>

bquan0 committed:
add docstrings and use Path instead of os

bquan0 committed:
raise exception due to invalid uv_map

Co-authored-by: connoramoreno <67344181+connoramoreno@users.noreply.github.com>

gonuke reviewed (commented):
I left one comment on the discussion of the `enumerate()` vs `range()`. Once we have resolution of that, I'm happy to merge.

bquan0 committed:
try enumerate solution for renaming collisions

Co-authored-by: connoramoreno <67344181+connoramoreno@users.noreply.github.com>

gonuke commented:
@bquan0 (since we don't have tests) can you confirm that this `enumerate()` change works as intended?

bquan0 commented:
@gonuke yep, looks like it worked. Thanks for the suggestion @connoramoreno!
